File: StepCounter.java
// StepCounter.java
import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.*;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.LocalDate;
import java.util.*;

public class StepCounter {
    private static final String DATA_FILE = "steps.json";
    private static final int DEFAULT_GOAL = 10000;
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static final Type DATA_TYPE = new TypeToken<Data>(){}.getType();

    @Parameter(names = "--add")
    private Integer addSteps;
    @Parameter(names = "--show")
    private boolean show;
    @Parameter(names = "--goal")
    private Integer goal;
    @Parameter(names = "--history")
    private boolean history;
    @Parameter(names = "--reset")
    private boolean reset;
    @Parameter(names = "--export-csv")
    private String exportCsv;

    static class Data {
        int goal = DEFAULT_GOAL;
        List<HistoryEntry> history = new ArrayList<>();
    }

    static class HistoryEntry {
        String date;
        int steps;
    }

    private Data data = new Data();

    private void load() {
        try {
            String json = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            data = GSON.fromJson(json, DATA_TYPE);
        } catch (Exception e) {
            data = new Data();
        }
        if (data.history == null) data.history = new ArrayList<>();
    }

    private void save() {
        try {
            Files.write(Paths.get(DATA_FILE), GSON.toJson(data).getBytes());
        } catch (IOException e) {
            System.err.println("Ошибка сохранения: " + e.getMessage());
        }
    }

    private HistoryEntry getToday() {
        String today = LocalDate.now().toString();
        for (HistoryEntry e : data.history) {
            if (e.date.equals(today)) return e;
        }
        return null;
    }

    private void addSteps(int steps) {
        HistoryEntry entry = getToday();
        if (entry != null) {
            entry.steps += steps;
        } else {
            entry = new HistoryEntry();
            entry.date = LocalDate.now().toString();
            entry.steps = steps;
            data.history.add(entry);
        }
        save();
        System.out.println("\u001B[32mДобавлено " + steps + " шагов. Всего сегодня: " + entry.steps + "\u001B[0m");
    }

    private void show() {
        HistoryEntry entry = getToday();
        int stepsToday = entry != null ? entry.steps : 0;
        int progress = Math.min(100, stepsToday * 100 / data.goal);
        int barLen = 20;
        int filled = progress * barLen / 100;
        String bar = "█".repeat(filled) + "░".repeat(barLen - filled);
        System.out.println("\u001B[36m📊 Сегодня: " + stepsToday + " шагов\u001B[0m");
        System.out.println("\u001B[33mЦель: " + data.goal + " шагов\u001B[0m");
        System.out.println("Прогресс: \u001B[32m" + bar + "\u001B[0m " + progress + "%");
        System.out.println("\u001B[35mИстория за 7 дней:\u001B[0m");
        LocalDate now = LocalDate.now();
        for (int i = 6; i >= 0; i--) {
            LocalDate d = now.minusDays(i);
            String dateStr = d.toString();
            HistoryEntry found = null;
            for (HistoryEntry e : data.history) {
                if (e.date.equals(dateStr)) { found = e; break; }
            }
            System.out.println("  " + dateStr + ": " + (found != null ? found.steps : 0));
        }
    }

    private void setGoal(int goal) {
        data.goal = goal;
        save();
        System.out.println("\u001B[32mЦель установлена: " + goal + " шагов\u001B[0m");
    }

    private void history() {
        System.out.println("\u001B[36mИстория (все записи):\u001B[0m");
        data.history.sort(Comparator.comparing(e -> e.date));
        for (HistoryEntry e : data.history) {
            System.out.println("  " + e.date + ": " + e.steps);
        }
    }

    private void reset() {
        data.history.clear();
        save();
        System.out.println("\u001B[32mВсе данные сброшены.\u001B[0m");
    }

    private void exportCsv(String filename) throws IOException {
        try (PrintWriter pw = new PrintWriter(new FileWriter(filename))) {
            pw.println("date,steps");
            for (HistoryEntry e : data.history) {
                pw.println(e.date + "," + e.steps);
            }
        }
        System.out.println("\u001B[32mЭкспортировано в " + filename + " (CSV)\u001B[0m");
    }

    public void run() throws Exception {
        load();
        if (addSteps != null) {
            addSteps(addSteps);
        } else if (show) {
            show();
        } else if (goal != null) {
            setGoal(goal);
        } else if (history) {
            history();
        } else if (reset) {
            reset();
        } else if (exportCsv != null) {
            exportCsv(exportCsv);
        } else {
            System.out.println("Используйте --help для справки.");
        }
    }

    public static void main(String[] args) throws Exception {
        StepCounter counter = new StepCounter();
        JCommander.newBuilder().addObject(counter).build().parse(args);
        counter.run();
    }
}

File: StepCounter.kt
// StepCounter.kt
import com.beust.jcommander.JCommander
import com.beust.jcommander.Parameter
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import java.io.File
import java.time.LocalDate

data class HistoryEntry(val date: String, val steps: Int)
data class Data(var goal: Int = 10000, val history: MutableList<HistoryEntry> = mutableListOf())

class StepCounter {
    @Parameter(names = ["--add"])
    private var addSteps: Int? = null

    @Parameter(names = ["--show"])
    private var show: Boolean = false

    @Parameter(names = ["--goal"])
    private var goal: Int? = null

    @Parameter(names = ["--history"])
    private var history: Boolean = false

    @Parameter(names = ["--reset"])
    private var reset: Boolean = false

    @Parameter(names = ["--export-csv"])
    private var exportCsv: String? = null

    private val dataFile = "steps.json"
    private val gson = GsonBuilder().setPrettyPrinting().create()
    private val type = object : TypeToken<Data>() {}.type
    private lateinit var data: Data

    init {
        load()
    }

    private fun load() {
        val f = File(dataFile)
        data = if (f.exists()) {
            try {
                val json = f.readText()
                gson.fromJson(json, type) ?: Data()
            } catch (e: Exception) {
                Data()
            }
        } else {
            Data()
        }
        if (data.history == null) data.history = mutableListOf()
    }

    private fun save() {
        File(dataFile).writeText(gson.toJson(data))
    }

    private fun getToday(): HistoryEntry? {
        val today = LocalDate.now().toString()
        return data.history.find { it.date == today }
    }

    private fun addSteps(steps: Int) {
        val entry = getToday()
        if (entry != null) {
            entry.steps += steps
        } else {
            data.history.add(HistoryEntry(LocalDate.now().toString(), steps))
        }
        save()
        val total = entry?.steps ?: steps
        println("\u001B[32mДобавлено $steps шагов. Всего сегодня: $total\u001B[0m")
    }

    private fun show() {
        val entry = getToday()
        val stepsToday = entry?.steps ?: 0
        val progress = (stepsToday * 100 / data.goal).coerceAtMost(100)
        val barLen = 20
        val filled = progress * barLen / 100
        val bar = "█".repeat(filled) + "░".repeat(barLen - filled)
        println("\u001B[36m📊 Сегодня: $stepsToday шагов\u001B[0m")
        println("\u001B[33mЦель: ${data.goal} шагов\u001B[0m")
        println("Прогресс: \u001B[32m$bar\u001B[0m $progress%")
        println("\u001B[35mИстория за 7 дней:\u001B[0m")
        for (i in 6 downTo 0) {
            val d = LocalDate.now().minusDays(i.toLong())
            val dateStr = d.toString()
            val found = data.history.find { it.date == dateStr }
            println("  $dateStr: ${found?.steps ?: 0}")
        }
    }

    private fun setGoal(goal: Int) {
        data.goal = goal
        save()
        println("\u001B[32mЦель установлена: $goal шагов\u001B[0m")
    }

    private fun history() {
        println("\u001B[36mИстория (все записи):\u001B[0m")
        data.history.sortedBy { it.date }.forEach {
            println("  ${it.date}: ${it.steps}")
        }
    }

    private fun reset() {
        data.history.clear()
        save()
        println("\u001B[32mВсе данные сброшены.\u001B[0m")
    }

    private fun exportCsv(filename: String) {
        File(filename).printWriter().use { pw ->
            pw.println("date,steps")
            data.history.forEach { pw.println("${it.date},${it.steps}") }
        }
        println("\u001B[32mЭкспортировано в $filename (CSV)\u001B[0m")
    }

    fun run() {
        when {
            addSteps != null -> addSteps(addSteps!!)
            show -> show()
            goal != null -> setGoal(goal!!)
            history -> history()
            reset -> reset()
            exportCsv != null -> exportCsv(exportCsv!!)
            else -> println("Используйте --help для справки.")
        }
    }
}

fun main(args: Array<String>) {
    val counter = StepCounter()
    JCommander.newBuilder().addObject(counter).build().parse(*args)
    counter.run()
}
